sdk/storage/Azure.Storage.Common/src/Shared/PartitionedUploader.cs

```diff
@@ -0,0 +1,607 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core.Pipeline;
+using Azure.Storage.Shared;
+
+namespace Azure.Storage
+{
+    internal class PartitionedUploader<TServiceSpecificArgs, TCompleteUploadReturn>
+    {
+        #region Definitions
+        // delegte for getting a partition from a stream based on the selected data management stragegy
+        private delegate Task<StreamSlice> GetNextStreamPartition(
+            Stream stream,
+            long minCount,
+            long maxCount,
+            long absolutePosition,
+            bool async,
+            CancellationToken cancellationToken);
+
+        // injected behaviors for services to use partitioned uploads
+        public delegate DiagnosticScope CreateScope(string operationName);
+        public delegate Task InitializeDestinationInternal(TServiceSpecificArgs args, bool async, CancellationToken cancellationToken);
+        public delegate Task<Response<TCompleteUploadReturn>> SingleUploadInternal(
+            Stream contentStream,
+            TServiceSpecificArgs args,
+            IProgress<long> progressHandler,
+            string operationName,
+            bool async,
+            CancellationToken cancellationToken);
+        public delegate Task UploadPartitionInternal(Stream contentStream,
+            long offset,
+            TServiceSpecificArgs args,
+            IProgress<long> progressHandler,
+            bool async,
+            CancellationToken cancellationToken);
+        public delegate Task<Response<TCompleteUploadReturn>> CommitPartitionedUploadInternal(
+            List<(long Offset, long Size)> partitions,
+            TServiceSpecificArgs args,
+            bool async,
+            CancellationToken cancellationToken);
+
+        public struct Behaviors
+        {
+            public InitializeDestinationInternal InitializeDestination { get; set; }
+            public SingleUploadInternal SingleUpload { get; set; }
+            public UploadPartitionInternal UploadPartition { get; set; }
+            public CommitPartitionedUploadInternal CommitPartitionedUpload { get; set; }
+            public CreateScope Scope { get; set; }
+        }
```

Comment:
@jaschrep-msft worked on this part.

Reply:
@kasobol-msft encouraged me to view them as individual, injectable behaviors instead of an object with a coherent functionality. The struct is just to make the constructor look smoother.

Reply:
I don't really understand the benefit.  We're not using any DI so this feels kind of extra?

Reply:
Composition based design is usually more flexible. Code reuse through inheritance leads into problems if the tree grows over time. So if there isn't good reason to create class family (there isn't really a hierarchy) we've chosen composition. [Here](https://en.wikipedia.org/wiki/Composition_over_inheritance) , [here](http://welllin.net/composition-over-inheritance/) and [here](http://blogs.perl.org/users/sid_burn/2014/03/inheritance-is-bad-code-reuse-part-1.html) are some good articles about this.

Reply:
I understand composition vs inheritance, but interfaces do not create a class hierarchy.  You can and should use interfaces to define the contract if you're designing for nontrivial composition.  You have a struct full of functions here which is effectively rolling your own version of a C# interface that's both slower and offers less compile time safety.  I commented specifically because that's not idiomatic.

Reply:
@tg-msft  gotcha. I was under impression we're talking about inheriting from PartitionedUploader and filling in gaps. 
Yeah, we had an interface at some point but it didn't look very cohesive, i.e. it was hard to name it right and justify why "Create Scope" sits together with "UploadPartition" for instance (the only reason they're together is because the uploader needs them, otherwise they don't have much in common). Therefore, passing delegates in a bag seemed more appealing.

Since it's all internal I guess we could go back to interface. However, I'd appreciate idea how to make that interface cohesive (maybe we need more than one).